src/Test/Perf/util/test_util.csx

```diff
@@ -12,125 +12,185 @@ using System.Threading;
-{
-    return Path.Combine(RoslynDirectory(), "src", "Test", "Perf");
-}
+    public string GetViBenchToJsonExeFilePath => Path.Combine(GetCPCDirectoryPath, "ViBenchToJson.exe");
+    
+    
+    
+    /// Downloads a zip from azure store and extracts it into
+    /// the ./temp directory.
+    ///
+    /// If this current version has already been downloaded
+    /// and extracted, do nothing.
+    public void DownloadProject(string name, int version)
+    {
+        var zipFileName = $"{name}.{version}.zip";
+        var zipPath = Path.Combine(MyTempDirectory, zipFileName);
+        // If we've already downloaded the zip, assume that it
+        // has been downloaded *and* extracted.
+        if (File.Exists(zipPath))
+        {
+            Log($"Didn't download and extract {zipFileName} because one already exists.");
+            return;
+        }
 
-static string BinDirectory()
-{
-    return Path.Combine(RoslynDirectory(), "Binaries");
-}
+        // Remove all .zip files that were downloaded before.
+        foreach (var path in Directory.EnumerateFiles(MyTempDirectory, $"{name}.*.zip"))
+        {
+            Log($"Removing old zip {path}");
+            File.Delete(path);
+        }
 
-static string BinDebugDirectory()
-{
-    return Path.Combine(BinDirectory(), "Debug");
+        // Download zip file to temp directory
+        var downloadTarget = $"https://dotnetci.blob.core.windows.net/roslyn-perf/{zipFileName}";
+        Log($"Downloading {downloadTarget}");
+        var client = new WebClient();
+        client.DownloadFile(downloadTarget, zipPath);
+        Log($"Done Downloading");
+
+        // Extract to temp directory
+        Log($"Extracting {zipPath} to {MyTempDirectory}");
+        ZipFile.ExtractToDirectory(zipPath, MyTempDirectory);
+        Log($"Done Extracting");
+    }
 }
 
-static string BinReleaseDirectory()
-{
-    return Path.Combine(BinDirectory(), "Release");
+abstract class PerfTest: RelativeDirectory {
+    private List<Tuple<int, string, object>> _metrics = new List<Tuple<int, string, object>>();
+    
+    public PerfTest([CallerFilePath] string workingFile = ""): base(workingFile) {}
+    
+    /// Reports a metric to be recorded in the performance monitor.
+    protected void Report(ReportKind reportKind, string description, object value)
+    {
+        _metrics.Add(Tuple.Create((int) reportKind, description, value));
+        Log(description + ": " + value.ToString());
+    }
+    
+    public abstract void Setup();
+    public abstract void Test();
+    public abstract int Iterations {get;}
 }
 
-static string DebugCscPath()
+static PerfTest[] TestThisPlease(params PerfTest[] tests) 
 {
-    return Path.Combine(BinDebugDirectory(), "csc.exe");
+    if (IsRunFromRunner()) 
+    {
+        return tests;
+    }
+    else 
+    {
+        foreach (var test in tests) 
+        {
+            for (int i = 0; i < test.Iterations; i++) 
+            {
+                test.Setup();
```

Comment:
I can imagine us wanting a Setup method that runs outside this loop and a separate Setup method that runs on each iteration. I can also imagine a cleanup step between each iteration.